src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,330 @@
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineControl.xaml
+    /// </summary>
+    internal partial class DocumentOutlineControl : UserControl, IOleCommandTarget
+    {
+        [MemberNotNullWhen(true, nameof(SymbolsTreeItemsSource))]
+        private bool SymbolTreeInitialized { get; set; }
+        private List<DocumentSymbolViewModel>? SymbolsTreeItemsSource { get; set; }
+
+        private ITextSnapshot? Snapshot { get; set; }
+        private IWpfTextView TextView { get; set; }
+
+        private readonly string? _filePath;
+        private ResettableDelay? _delay;
+        private readonly IAsynchronousOperationListener _asyncListener;
+
+        public DocumentOutlineControl(IWpfTextView textView, ITextBuffer textBuffer, ILanguageServiceBroker2 languageServiceBroker, IThreadingContext threadingContext, IAsynchronousOperationListener asyncListener)
+        {
+            InitializeComponent();
+
+            _asyncListener = asyncListener;
+            TextView = textView;
+            _filePath = GetFilePath(textView);
+            Snapshot = textBuffer.CurrentSnapshot;
+            SymbolTreeInitialized = false;
+
+            threadingContext.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await UpdateAsync().ConfigureAwait(false);
+            }).FileAndForget("Document Outline: Active Document Changed");
+
+            TextView.Caret.PositionChanged += FollowCursor;
+            TextView.TextBuffer.Changed += TextBuffer_Changed;
+
+            void TextBuffer_Changed(object sender, TextContentChangedEventArgs e)
+                => EnqueueUpdate();
+
+            void EnqueueUpdate()
+            {
+                const int UpdateDelay = 500;
+                var delay = _delay;
+                if (delay == null)
+                {
+                    var newDelay = new ResettableDelay(UpdateDelay, _asyncListener);
+                    if (Interlocked.CompareExchange(ref _delay, newDelay, null) == null)
+                    {
+                        var asyncToken = _asyncListener.BeginAsyncOperation("Updating Document Outline");
+                        newDelay.Task.SafeContinueWithFromAsync(_ => UpdateAsync(), CancellationToken.None, TaskScheduler.Default)
+                            .SafeContinueWith(_ => _delay = null, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default)
+                            .CompletesAsyncOperation(asyncToken);
+                    }
+
+                    return;
+                }
+
+                delay.Reset();
+            }
+
+            async Task UpdateAsync()
+            {
+                var response = await DocumentSymbolsRequestAsync(textBuffer, languageServiceBroker).ConfigureAwait(false);
```

Comment:
cosnider moving *explicitly* off the UI thread so that getting this result doesn't occupy UI time.

Reply:
Done